drivers: uart: Add low power uart module (#2591)
nrfconnect/sdk-nrf: doc/nrf/drivers/uart_nrf_lpuart.rst

Additionally, the standard UART configuration must have flow control and flow control pins disabled.

See the following configuration example:

.. code-block:: DTS

	&uart1 {
		compatible = "nordic,nrf-uarte";
		status = "okay";
		rx-pin = <44>;
		tx-pin = <45>;
		baudrate = <1000000>;
		/delete-property/ rts-pin;
		/delete-property/ cts-pin;
		/delete-property/ hw-flow-control;

		lpuart: nrf-sw-lpuart {
			compatible = "nordic,nrf-sw-lpuart";
			status = "okay";
			label = "LPUART";
			req-pin = <46>;
			rdy-pin = <47>;
		};
	};

The low power UART configuration includes:

* :option:`CONFIG_NRF_LPUART_UARTE_INSTANCE`: the UART instance choice
* :option:`CONFIG_NRF_LPUART_REQ_PIN`: the Request line pin
* :option:`CONFIG_NRF_LPUART_RDY_PIN`: the Ready line pin

Usage
*****

You can access and control the low power UART using the Asynchronous UART API.

Data is sent using :cpp:func:`uart_tx`.
The transfer will timeout if the receiver does not acknowledge its readiness.

The receiver is enabled by calling :cpp:func:`uart_rx_enable`.
After that call, the receiver is set up and set to idle (low power) state.

> See :ref:`lpuart_sample` for an implementation of this driver.

Reviewer: ```suggestion
See :ref:`lpuart_sample` sample for an implementation of this driver.
```

.. _uart_nrf_lpuart:

Low power UART driver
#####################

> The low power UART driver is a driver which implements the standard asynchronous UART API: :option:`CONFIG_UART_ASYNC_API`.

Reviewer 2: fixed

.. _uart_nrf_lpuart:

Low power UART driver
#####################

The low power UART driver is a driver which implements the standard asynchronous UART API: :option:`CONFIG_UART_ASYNC_API`.

> It uses a two control lines protocol, instead of standard hardware flow control lines, to allow disabling the UART receiver during the idle period.

Reviewer 2: fixed

> This allows to achieve a low power consumption, as you can shut down the high frequency clock when the UART is in idle.

Reviewer 2: fixed

Control protocol
****************

You can use the protocol for duplex transmission between two devices.

The control protocol uses two additional lines alongside the standard TX and RX lines:

> * the *Request* line (REQ)

Reviewer 2: fixed

> * the *Ready* line (RDY)

Reviewer 2: fixed

The REQ line of the first device is connected with the RDY line of the second device.
It is configured as output and set to low.

The RDY line is configured as input without pull up and the interrupt is configured to detect transition from low to high state.

> Then:

Reviewer 2: Fixed for now, waiting for feedback from @nordic-krch

> #. The transmitter initiate the transmission by calling :cpp:func:`uart_tx`.

Reviewer 2: fixed

#. The driver reconfigures the REQ line to input with pull up and enables the detection of high to low transition.

>    The line is set to high, due to the pull up register, and that triggers an interrupt on the RDY line.

Reviewer: ```suggestion
   The line is set to high due to the pull-up register, and that triggers an interrupt on the RDY line.
```

Reviewer 2: fixed

#. On that interrupt, the driver starts the UART receiver.
#. When the receiver is ready, the driver reconfigures the RDY line to output low.
#. Then, the driver reconfigures the RDY line to input with pull up and enables the interrupt on detection of high to low transition.
#. This sequence results in a short pulse on the line, as the line goes low and high.
#. The initiator detects the pulse and starts the standard UART transfer.

> #. When the transfer is completed, the driver reconfigure the REQ line to the initial state: output set to low.

Reviewer: ```suggestion
#. When the transfer is completed, the driver reconfigures the REQ line to the initial state: output set to low.
```

Reviewer 2: fixed

   This results in a line state change.
#. As the line goes low, the receiver detects the change.
   This indicates that the UART receiver can be stopped.

Once the receiver acknowledges the transfer, it must be capable of receiving the whole transfer until the REQ line goes down.

This requirement can be fullfilled in two ways:

> * By providing a buffer of the size of the maximum packet length

Reviewer: ```suggestion
* By providing a buffer of the size of the maximum packet length.
```

Reviewer 2: fixed

* By continuously responding to the RX buffer request event.
  The latency of the event handling must be taken into account in that case.

>   For example, a flash page erase on some devices may have a significant impact.

Reviewer: ```suggestion
  For example, a flash page erase on some devices might have a significant impact.
```
(Avoid using 'may')

Reviewer 2: fixed

Configuration
*************

> The low power UART driver is built on top of the standard UART driver extended by control lines protocol.

Reviewer: ```suggestion
The low power UART driver is built on top of the standard UART driver extended with the control lines protocol.
```

Reviewer 2: fixed

> It is configured in the device tree as the child node of the UART instance which extends standard UART configuration with REQ and RDY lines.

Reviewer: ```suggestion
It is configured in the device tree as a child node of the UART instance that extends standard UART configuration with REQ and RDY lines.
```

Reviewer 2: fixed

Additionally, the standard UART configuration must have flow control and flow control pins disabled.

> See the following example:

Reviewer: Please state what this example shows.

Reviewer 2: fixed, @nordic-krch please check this one

.. code-block:: DTS

	&uart1 {
		compatible = "nordic,nrf-uarte";
		status = "okay";
		rx-pin = <44>;
		tx-pin = <45>;
		baudrate = <1000000>;
		/delete-property/ rts-pin;
		/delete-property/ cts-pin;
		/delete-property/ hw-flow-control;
	
		lpuart: nrf-sw-lpuart {
			compatible = "nordic,nrf-sw-lpuart";
			status = "okay";
			label = "LPUART";
			req-pin = <46>;
			rdy-pin = <47>;
		};
	};

The low power UART configuration includes:

* :option:`CONFIG_NRF_LPUART_UARTE_INSTANCE`: the UART instance choice
* :option:`CONFIG_NRF_LPUART_REQ_PIN`: the Request line pin
* :option:`CONFIG_NRF_LPUART_RDY_PIN`: the Ready line pin

Usage
*****

> You can access and control the low power UART using the Asynchronous UART API.

Reviewer: We should add a link to this API. If I understand correctly, it's this one:
https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/zephyr/reference/peripherals/uart.html
@nordic-krch  please confirm
If yes, then it should be :ref:\`zephyr:uart_api\`

Reviewer 2: waiting on this one for confirmation from @nordic-krch
